### PR TITLE
SONAR-10261 Branches should not be able to override project name, description and links

### PR DESCRIPTION
**Pull request by sns-seb**

### CODE REVIEW

File: server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/analysis/Project.java [hunk 1]
```diff
@@ -77,15 +86,9 @@ public String toString() {
     sb.append("uuid='").append(uuid).append('\'');
     sb.append(", key='").append(key).append('\'');
     sb.append(", name='").append(name).append('\'');
+    sb.append(", description=").append(description == null ? null : '\'' + description + '\'');
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/analysis/Project.java]: Code Smell Code Smell: Add parentheses to make the operator precedence explicit. ([squid:S864](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS864))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarqube%3Asonarqube&branch=feature%2Fsl%2FSONAR-10261%2Fbranch_analysis_does_not_update_name_description_nor_links&issues=AWH8Yt3ja5Ltb7kAT2N8&open=AWH8Yt3ja5Ltb7kAT2N8)

[ISSUE_KEY]: AWH8Yt3ja5Ltb7kAT2N8

File: server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/step/PersistProjectLinksStep.java [hunk 2]
```diff
@@ -51,7 +53,8 @@
     ComponentLinkType.CI, ProjectLinkDto.TYPE_CI,
     ComponentLinkType.ISSUE, ProjectLinkDto.TYPE_ISSUE_TRACKER);
 
-  public PersistProjectLinksStep(DbClient dbClient, TreeRootHolder treeRootHolder, BatchReportReader reportReader, UuidFactory uuidFactory) {
+  public PersistProjectLinksStep(AnalysisMetadataHolder analysisMetadataHolder, DbClient dbClient, TreeRootHolder treeRootHolder, BatchReportReader reportReader, UuidFactory uuidFactory) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/step/PersistProjectLinksStep.java]: Code Smell Code Smell: Split this 188 characters long line (which is greater than 180 authorized). ([squid:S00103](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS00103))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarqube%3Asonarqube&branch=feature%2Fsl%2FSONAR-10261%2Fbranch_analysis_does_not_update_name_description_nor_links&issues=AWH_oDnB9MctqhxriTqq&open=AWH_oDnB9MctqhxriTqq)

[ISSUE_KEY]: AWH_oDnB9MctqhxriTqq

File: server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/analysis/Project.java [hunk 3]
```diff
@@ -77,15 +86,13 @@ public String toString() {
...
     sb.append(", name='").append(name).append('\'');
+    sb.append(", description=").append(toString(this.description));
     sb.append('}');
     return sb.toString();
   }
 
-  public static Project copyOf(Component component) {
-    return new Project(component.getUuid(), component.getKey(), component.getName());
+  private static String toString(@Nullable String s) {
+    return s == null ? null : '\'' + s + '\'';
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/analysis/Project.java]: Code Smell Code Smell: Add parentheses to make the operator precedence explicit. ([squid:S864](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS864))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarqube%3Asonarqube&branch=feature%2Fsl%2FSONAR-10261%2Fbranch_analysis_does_not_update_name_description_nor_links&issues=AWH_t2Bo9MctqhxriUSu&open=AWH_t2Bo9MctqhxriUSu)

[ISSUE_KEY]: AWH_t2Bo9MctqhxriUSu

File: server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/analysis/Project.java [hunk 4]
```diff
@@ -77,15 +86,16 @@ public String toString() {
...
-  public static Project copyOf(Component component) {
-    return new Project(component.getUuid(), component.getKey(), component.getName());
+  private static String toString(@Nullable String s) {
+    if (s == null) {
+      return null;
+    }
+    return '\'' + s + '\'';
   }
 
-  public static Project copyOf(ComponentDto component) {
```
Comment (janos-ss) [inline server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/analysis/Project.java]: What was wrong with these `copyOf` methods?

File: server/sonar-server/src/test/java/org/sonar/server/computation/task/projectanalysis/analysis/ProjectTest.java [hunk 5]
```diff
@@ -20,46 +20,58 @@
...
+    assertThat(project1.getUuid()).isEqualTo(project2.getUuid()).isEqualTo("U1");
+    assertThat(project1.getKey()).isEqualTo(project2.getKey()).isEqualTo("K1");
+    assertThat(project1.getName()).isEqualTo(project2.getName()).isEqualTo("N1");
+    assertThat(project1.getDescription()).isEqualTo(project2.getDescription()).isNull();
+
+    assertThat(project1.toString())
+      .isEqualTo(project2.toString())
+      .isEqualTo("Project{uuid='U1', key='K1', name='N1', description=null}");
+  }
+  @Test
```
Comment (janos-ss) [inline server/sonar-server/src/test/java/org/sonar/server/computation/task/projectanalysis/analysis/ProjectTest.java]: An empty line between test cases would be nice :-)

File: server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/analysis/Project.java (hunk 4 above)
Comment (sns-seb) [reply server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/analysis/Project.java]: they were used only in unit tests and I'm pretty sure introduced in production code for this only purpose => delete

**Pull request merged**